### PR TITLE
DOC: Add and harmonise docstrings and type annotations

### PR DESCRIPTION
**Pull request by Patol75**
Given the current effort to revamp the G-ADOPT website, this PR attempts to improve the codebase documentation. Comments and suggestions are welcome!

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -7,10 +7,10 @@
 # - E226 (missing whitespace around arithmetic operator)
 # - E741 (ambiguous variable name)
 # - E701 (multiple statements on one line (colon))
-# - W504 (line break after binary operator)
+# - W503 (line break occurred before a binary operator)
 
 ignore =
-    E501,F403,F405,E226,E741,E701,W504
+    E501,F403,F405,E226,E741,E701,W503
```
[the remark this answers is not in the document]
Comment (Patol75) [reply setup.cfg:13]: Can we have both W503 and W504 then?

File: gadopt/utility.py [hunk 2]
```diff
@@ -1,22 +1,43 @@
...
+)
+from firedrake.__future__ import Interpolator
 from firedrake.petsc import PETSc
 from mpi4py import MPI
-import numpy as np
-import logging
-from logging import DEBUG, INFO, WARNING, ERROR, CRITICAL  # NOQA
-import os
 from scipy.linalg import solveh_banded
+from ufl.corealg.traversal import traverse_unique_terminals
```
[the remark this answers is not in the document]
Comment (Patol75) [reply gadopt/utility.py]: Sorry, I did not get to reverse all of that.

File: gadopt/energy_solver.py [hunk 3]
```diff
@@ -1,8 +1,33 @@
...
+from .utility import (
+    DEBUG,
+    INFO,
+    absv,
+    ensure_constant,
+    is_continuous,
+    log,
+    log_level,
+    su_nubar,
+)
```
[the remark this answers is not in the document]
Comment (Patol75) [reply gadopt/energy_solver.py]: I like `isort`, but this reformatting is indeed ugly. I will put up a draft PR to show how we could deal with `import`s in `gadopt` source files.

File: gadopt/diagnostics.py
[the remark this answers is not in the document]
Comment (Patol75) [reply gadopt/diagnostics.py:1]: Good catch!

File: gadopt/equations.py [hunk 4]
```diff
@@ -1,44 +1,69 @@
...
-        :arg test_space: the test functionspace
-        :arg trial_space: The trial functionspace
-        test and trial space are only used to determine the the discretisation that's used (ufl_element)
-        not what test and trial functions are actually used (these are provided seperately in residual())
-        :arg quad_degree: quadrature degree, default is 2*p+1 where p is the polynomial degree of trial_space
-        any further keywords arguments are passed on to the individual terms
+    def __init__(
+        self,
+        test_space: firedrake.functionspaceimpl.WithGeometry,
+        trial_space: firedrake.functionspaceimpl.WithGeometry,
```
[the remark this answers is not in the document]
Comment (Patol75) [reply gadopt/equations.py:36]: I had the same feeling!

File: gadopt/momentum_equation.py [hunk 5]
```diff
@@ -81,12 +122,12 @@ def residual(self, test, trial, trial_lagged, fields, bcs):
         # instead of maximum over two adjacent cells + and -, we just sum (which is 2*avg())
         # and the for internal facets we have an extra 0.5:
         # WEIRDNESS: avg(1/CellVolume(mesh)) crashes TSFC - whereas it works in scalar diffusion! - instead just writing out explicitly
-        sigma *= FacetArea(self.mesh)*(1/CellVolume(self.mesh)('-') + 1/CellVolume(self.mesh)('+'))/2
+        sigma *= fd.FacetArea(self.mesh)*(1/fd.CellVolume(self.mesh)('-') + 1/fd.CellVolume(self.mesh)('+'))/2
```
[the remark this answers is not in the document]
Comment (Patol75) [reply gadopt/momentum_equation.py]: I discussed that with Steph and Will the other night. I agree we need to make things consistent. I will have a try.

File: gadopt/momentum_equation.py [hunk 6]
```diff
@@ -1,51 +1,92 @@
...
+          trial_lagged:
+            Firedrake trial function from the previous time step.
+          fields:
+            Dictionary of physical fields from the simulation's state.
+          bcs:
+            Dictionary of identifier-value pairs specifying boundary conditions.
+
+        Returns:
+          A UFL expression for the term's contribution to the finite element residual.
+        """
```
[the remark this answers is not in the document]
Comment (Patol75) [reply gadopt/momentum_equation.py]: This is very true. `typing` has the `@overriden`, which is explicit and thus very useful, but this is only available from Python 3.12.

**Review by drhodrid — APPROVED** [review]
I think this looks REALLY good. My thoughts are that we try to get it in soon -- there are a substantial amount of changes in this PR and the webpage will look more complete with it in -- and then we add subsequent commits that address missing documentation. Immediate points of weakness for future PR's are:
1. Inverse.py - which @sghelichkhani  can probably address in a subsequent pull request.
2. Preconditioners.py - which @stephankramer is probably best placed to address.

**Comment by drhodrid** [comment]
@angus-g - happy to merge this and check how it appears on the new webpage afterwards?

**Comment by angus-g** [comment]
Yep, let's do that!

**Pull request merged**